### PR TITLE
refactor(x/data): remove unique constraint on resolver url

### PR DESCRIPTION
**Pull request by ryanchristo**
## Description

Closes: #1111

This pull request removes the unique constraint on url and updates query resolver to take the resolver id rather than url.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by aleem1314 — APPROVED** [review]
lgtm

File: x/data/client/testsuite/grpc.go [hunk 1]
```diff
@@ -474,22 +472,22 @@ func (s *IntegrationTestSuite) TestQueryResolverInfo() {
...
 			"not found",
 			0,
 		},
 		{
 			"valid request",
-			fmt.Sprintf("%s/regen/data/v1/resolver?url=%s", val.APIAddress, url),
+			fmt.Sprintf("%s/regen/data/v1/resolver?id=%d", val.APIAddress, s.resolverID),
 			false,
 			"",
 			2,
```
Comment (clevinson) [inline x/data/client/testsuite/grpc.go:486]: Doesn't look like `expItems` is ever checked. Should we remove it from the table?

Also - I don't think we're actually able to query resolver with pagination, so maybe we should drop this last test case entirely?

File: x/data/server/features/define_resolver.feature [hunk 2]
```diff
@@ -1,15 +1,12 @@
 Feature: Define Resolver
 
-  Background:
-    Given the resolver url "https://foo.bar"
+  Rule: the resolver is defined with no unique constraints on the URL
 
-  Rule: the resolver is defined if the resolver url is unique
+    Scenario: the url is unique
```
Comment (clevinson) [inline x/data/server/features/define_resolver.feature:5]:
```suggestion
    Scenario: a resolver for a new URL is registered
```

File: x/data/server/features/define_resolver.feature [hunk 3]
```diff
@@ -1,15 +1,12 @@
...
 
-    Scenario: the resolver has not been defined
-      When alice attempts to define the resolver
-      Then the resolver exists and alice is the manager
-
-    Scenario: the resolver has already been defined
-      Given alice has defined the resolver
-      When alice attempts to define the resolver
-      Then expect the error "resolver URL already exists"
+    Scenario: the url is not unique
```
Comment (clevinson) [inline x/data/server/features/define_resolver.feature:9]:
```suggestion
    Scenario: A resolver is registered with a URL that is already registered
```

**Review by clevinson — APPROVED** [review]
🙏 thank u @ryanchristo !

Just dropped at note in https://github.com/regen-network/regen-ledger/issues/1111#issuecomment-1137861872 echo'ing that I think querying for resolvers by URL with pagination (returning all resolvers with their ID & manager) would be useful.  Potentially in a follow-up.

File: x/data/server/features/define_resolver.feature (hunk 2 above)
Comment (ryanchristo) [reply x/data/server/features/define_resolver.feature:5]: In general, I was thinking the scenario would be short and to the point and the rule would be more descriptive. Given the rule is "the resolver is defined with no unique constraints on the URL", the two scenarios are "the url is unique" and "the url is not unique". This is also how the BDD books write scenarios. For example,

```
Rule: Any visitor to the website can place a customer-collection order

  Scenario: Unauthenticated customer chooses to collect order
    Given the customer is unauthenticated
    When they choose to collect their order
    Then they should be asked to supply contact details

  Scenario: Authenticated customer chooses to collect order
    Given the customer is authenticated
    When they choose to collect their order
    Then they should be asked to confirm contact details
```

Comment (ryanchristo) [reply x/data/server/features/define_resolver.feature:5]: This also produces the test logs:

```
=== RUN   TestDefineResolver
=== RUN   TestDefineResolver/Define_Resolver
=== RUN   TestDefineResolver/Define_Resolver/a_resolver_for_a_new_url_is_registered
=== RUN   TestDefineResolver/Define_Resolver/a_resolver_is_registered_with_a_url_that_is_already_registered
```

Versus the current test logs:

```
=== RUN   TestDefineResolver
=== RUN   TestDefineResolver/Define_Resolver
=== RUN   TestDefineResolver/Define_Resolver/the_url_is_unique
=== RUN   TestDefineResolver/Define_Resolver/the_url_is_not_unique
```

Comment (ryanchristo) [reply x/data/server/features/define_resolver.feature:5]: We are also "defining" a resolver here, not "registering" data to the resolver.

File: x/data/client/testsuite/grpc.go (hunk 1 above)
Comment (ryanchristo) [reply x/data/client/testsuite/grpc.go]: Good catch. Looks like some old copy pasta that went unnoticed.

Comment (ryanchristo) [reply x/data/client/testsuite/grpc.go]: https://github.com/regen-network/regen-ledger/pull/1128/commits/56f1b975eeab03dc2338fe4a9b3de0e53fa24c0f

**Comment by ryanchristo** [comment]
@clevinson Would be happy to further discuss best format for scenarios and update throughout but the current format is aligned with how I've been writing scenarios elsewhere and I think that if we include a descriptive rule, scenarios should be concise and provide only the necessary information, which makes reading through the test logs easier when debugging failing scenarios.

**Pull request merged**